File: rebuild/v2/client/src/cameraWheel.ts
import { wrap32 } from "./tileWheel32";

const METATILE_STEP_SUBTILES = 2;
const REDRAW_OPPOSITE_EDGE_OFFSET = 28;

export type CameraWheelState = {
  cameraPosX: number;
  cameraPosY: number;
  xTileOffset: number;
  yTileOffset: number;
};

export type StripRedrawSample = {
  destX: number;
  destY: number;
  worldX: number;
  worldY: number;
};

export function applyMetatileStep(state: CameraWheelState, deltaX: number, deltaY: number): StripRedrawSample[] {
  if (deltaX === 0 && deltaY === 0) {
    return [];
  }

  // Emerald order: update camera position first.
  state.cameraPosX += deltaX;
  state.cameraPosY += deltaY;

  // Then AddCameraTileOffset(delta * 2) in subtile units.
  state.xTileOffset = wrap32(state.xTileOffset + deltaX * METATILE_STEP_SUBTILES);
  state.yTileOffset = wrap32(state.yTileOffset + deltaY * METATILE_STEP_SUBTILES);

  const strip: StripRedrawSample[] = [];
  if (deltaX !== 0) {
    const destX = deltaX > 0 ? wrap32(state.xTileOffset + REDRAW_OPPOSITE_EDGE_OFFSET) : state.xTileOffset;
    const worldX = deltaX > 0 ? state.cameraPosX + 14 : state.cameraPosX;
    for (let i = 0; i <= 30; i += 2) {
      strip.push({
        destX,
        destY: wrap32(state.yTileOffset + i),
        worldX,
        worldY: state.cameraPosY + i / METATILE_STEP_SUBTILES
      });
    }
  }

  if (deltaY !== 0) {
    const destY = deltaY > 0 ? wrap32(state.yTileOffset + REDRAW_OPPOSITE_EDGE_OFFSET) : state.yTileOffset;
    const worldY = deltaY > 0 ? state.cameraPosY + 14 : state.cameraPosY;
    for (let i = 0; i <= 30; i += 2) {
      strip.push({
        destX: wrap32(state.xTileOffset + i),
        destY,
        worldX: state.cameraPosX + i / METATILE_STEP_SUBTILES,
        worldY
      });
    }
  }

  return strip;
}

export function runCameraWheelFixtures(): void {
  const north: CameraWheelState = { cameraPosX: 100, cameraPosY: 200, xTileOffset: 6, yTileOffset: 10 };
  const northStrip = applyMetatileStep(north, 0, 1);
  expectEqual(north.yTileOffset, 12, "north.yTileOffset");
  expectEqual(northStrip.length, 16, "north.strip.length");
  expectSample(northStrip[0], { destX: 6, destY: 8, worldX: 100, worldY: 215 }, "north.strip.first");
  expectSample(northStrip[15], { destX: 4, destY: 8, worldX: 115, worldY: 215 }, "north.strip.last");

  const west: CameraWheelState = { cameraPosX: 40, cameraPosY: 20, xTileOffset: 30, yTileOffset: 0 };
  const westStrip = applyMetatileStep(west, 1, 0);
  expectEqual(west.xTileOffset, 0, "west.xTileOffset");
  expectSample(westStrip[0], { destX: 28, destY: 0, worldX: 55, worldY: 20 }, "west.strip.first");
  expectSample(westStrip[15], { destX: 28, destY: 30, worldX: 55, worldY: 35 }, "west.strip.last");

  const wrap: CameraWheelState = { cameraPosX: 4, cameraPosY: 5, xTileOffset: 31, yTileOffset: 0 };
  applyMetatileStep(wrap, 1, -1);
  expectEqual(wrap.xTileOffset, 1, "wrap.xTileOffset");
  expectEqual(wrap.yTileOffset, 30, "wrap.yTileOffset");
}

function expectEqual(actual: number, expected: number, label: string): void {
  if (actual !== expected) {
    throw new Error(`[camera-wheel-fixture] ${label}: expected ${expected}, got ${actual}`);
  }
}

function expectSample(actual: StripRedrawSample, expected: StripRedrawSample, label: string): void {
  if (
    actual.destX !== expected.destX ||
    actual.destY !== expected.destY ||
    actual.worldX !== expected.worldX ||
    actual.worldY !== expected.worldY
  ) {
    throw new Error(`[camera-wheel-fixture] ${label}: expected ${JSON.stringify(expected)}, got ${JSON.stringify(actual)}`);
  }
}

File: rebuild/v2/server/src/camera_wheel.rs
use crate::map_runtime::MAP_OFFSET;

pub const WHEEL_SUBTILE_SIZE: i32 = 32;
pub const METATILE_STEP_SUBTILES: i32 = 2;
const REDRAW_SLICE_STRIDE: i32 = 2;
const REDRAW_SLICE_SPAN: i32 = 32;
const REDRAW_OPPOSITE_EDGE_OFFSET: i32 = 28;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct CameraWheelState {
    pub camera_pos_x: i32,
    pub camera_pos_y: i32,
    pub x_tile_offset: i32,
    pub y_tile_offset: i32,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct StripRedrawSample {
    pub dest_x: i32,
    pub dest_y: i32,
    pub world_x: i32,
    pub world_y: i32,
}

impl CameraWheelState {
    pub fn from_camera_runtime(camera_runtime_x: i32, camera_runtime_y: i32) -> Self {
        let camera_pos_x = camera_runtime_x - MAP_OFFSET as i32;
        let camera_pos_y = camera_runtime_y - MAP_OFFSET as i32;
        Self {
            camera_pos_x,
            camera_pos_y,
            x_tile_offset: wrap32(camera_pos_x * METATILE_STEP_SUBTILES),
            y_tile_offset: wrap32(camera_pos_y * METATILE_STEP_SUBTILES),
        }
    }

    pub fn apply_metatile_step(&mut self, delta_x: i32, delta_y: i32) -> Vec<StripRedrawSample> {
        if delta_x == 0 && delta_y == 0 {
            return Vec::new();
        }

        // Emerald order: update camera map position first.
        self.camera_pos_x += delta_x;
        self.camera_pos_y += delta_y;

        // Then AddCameraTileOffset(delta*2) in subtile units.
        self.x_tile_offset = wrap32(self.x_tile_offset + delta_x * METATILE_STEP_SUBTILES);
        self.y_tile_offset = wrap32(self.y_tile_offset + delta_y * METATILE_STEP_SUBTILES);

        // Then redraw strips using updated offsets.
        let mut strips = Vec::new();
        if delta_x != 0 {
            strips.extend(self.redraw_vertical_strip(delta_x));
        }
        if delta_y != 0 {
            strips.extend(self.redraw_horizontal_strip(delta_y));
        }
        strips
    }

    fn redraw_horizontal_strip(&self, delta_y: i32) -> Vec<StripRedrawSample> {
        let mut strip = Vec::with_capacity((REDRAW_SLICE_SPAN / REDRAW_SLICE_STRIDE) as usize);
        let dest_y = if delta_y > 0 {
            wrap32(self.y_tile_offset + REDRAW_OPPOSITE_EDGE_OFFSET)
        } else {
            self.y_tile_offset
        };
        let world_y = if delta_y > 0 {
            self.camera_pos_y + 14
        } else {
            self.camera_pos_y
        };

        for i in (0..REDRAW_SLICE_SPAN).step_by(REDRAW_SLICE_STRIDE as usize) {
            strip.push(StripRedrawSample {
                dest_x: wrap32(self.x_tile_offset + i),
                dest_y,
                world_x: self.camera_pos_x + i / METATILE_STEP_SUBTILES,
                world_y,
            });
        }

        strip
    }

    fn redraw_vertical_strip(&self, delta_x: i32) -> Vec<StripRedrawSample> {
        let mut strip = Vec::with_capacity((REDRAW_SLICE_SPAN / REDRAW_SLICE_STRIDE) as usize);
        let dest_x = if delta_x > 0 {
            wrap32(self.x_tile_offset + REDRAW_OPPOSITE_EDGE_OFFSET)
        } else {
            self.x_tile_offset
        };
        let world_x = if delta_x > 0 {
            self.camera_pos_x + 14
        } else {
            self.camera_pos_x
        };

        for i in (0..REDRAW_SLICE_SPAN).step_by(REDRAW_SLICE_STRIDE as usize) {
            strip.push(StripRedrawSample {
                dest_x,
                dest_y: wrap32(self.y_tile_offset + i),
                world_x,
                world_y: self.camera_pos_y + i / METATILE_STEP_SUBTILES,
            });
        }

        strip
    }
}

fn wrap32(value: i32) -> i32 {
    value.rem_euclid(WHEEL_SUBTILE_SIZE)
}

#[cfg(test)]
mod tests {
    use super::{CameraWheelState, StripRedrawSample};

    #[test]
    fn one_step_north_uses_updated_offsets_and_plus_28_destination_row() {
        let mut state = CameraWheelState {
            camera_pos_x: 100,
            camera_pos_y: 200,
            x_tile_offset: 6,
            y_tile_offset: 10,
        };

        let strip = state.apply_metatile_step(0, 1);

        assert_eq!(state.x_tile_offset, 6);
        assert_eq!(state.y_tile_offset, 12);
        assert_eq!(strip.len(), 16);
        assert_eq!(strip[0], sample(6, 8, 100, 215));
        assert_eq!(strip[15], sample(4, 8, 115, 215));
    }

    #[test]
    fn one_step_south_redraws_at_current_y_offset() {
        let mut state = CameraWheelState {
            camera_pos_x: 2,
            camera_pos_y: 3,
            x_tile_offset: 30,
            y_tile_offset: 2,
        };

        let strip = state.apply_metatile_step(0, -1);

        assert_eq!(state.y_tile_offset, 0);
        assert_eq!(strip[0], sample(30, 0, 2, 2));
        assert_eq!(strip[1], sample(0, 0, 3, 2));
    }

    #[test]
    fn one_step_east_wrap_case_uses_current_x_offset() {
        let mut state = CameraWheelState {
            camera_pos_x: 40,
            camera_pos_y: 20,
            x_tile_offset: 1,
            y_tile_offset: 31,
        };

        let strip = state.apply_metatile_step(-1, 0);

        assert_eq!(state.x_tile_offset, 31);
        assert_eq!(strip[0], sample(31, 31, 39, 20));
        assert_eq!(strip[1], sample(31, 1, 39, 21));
    }

    #[test]
    fn one_step_west_wrap_case_uses_plus_28_destination_column() {
        let mut state = CameraWheelState {
            camera_pos_x: 40,
            camera_pos_y: 20,
            x_tile_offset: 30,
            y_tile_offset: 0,
        };

        let strip = state.apply_metatile_step(1, 0);

        assert_eq!(state.x_tile_offset, 0);
        assert_eq!(strip[0], sample(28, 0, 55, 20));
        assert_eq!(strip[15], sample(28, 30, 55, 35));
    }

    #[test]
    fn diagonal_step_emits_independent_axis_strips() {
        let mut state = CameraWheelState {
            camera_pos_x: 12,
            camera_pos_y: 8,
            x_tile_offset: 28,
            y_tile_offset: 30,
        };

        let strips = state.apply_metatile_step(1, -1);

        assert_eq!(state.x_tile_offset, 30);
        assert_eq!(state.y_tile_offset, 28);
        assert_eq!(strips.len(), 32);
        assert_eq!(strips[0], sample(26, 28, 27, 7));
        assert_eq!(strips[16], sample(30, 28, 13, 7));
    }

    fn sample(dest_x: i32, dest_y: i32, world_x: i32, world_y: i32) -> StripRedrawSample {
        StripRedrawSample {
            dest_x,
            dest_y,
            world_x,
            world_y,
        }
    }
}
